File: msf/broadcast_test.go
package msf

import (
	"encoding/json"
	"testing"

	"github.com/okdaichi/gomoqt/moqt"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBroadcastRegisterTrack_UpdatesCatalogAndHandler(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)

	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })
	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.NoError(t, err)

	catalog := broadcast.Catalog()
	require.Len(t, catalog.Tracks, 1)
	assert.Equal(t, "video", catalog.Tracks[0].Name)
	tw := &moqt.TrackWriter{TrackName: "video"}
	handler.On("ServeTrack", tw).Return().Once()
	broadcast.Handler("video").ServeTrack(tw)
}

func TestBroadcastHandler_ResolvesCatalogAndTrackHandlers(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)

	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })
	err = broadcast.RegisterTrack(Track{
		Name:      "audio",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.NoError(t, err)

	assert.NotNil(t, broadcast.Handler(broadcast.CatalogTrackName()))
	tw := &moqt.TrackWriter{TrackName: "audio"}
	handler.On("ServeTrack", tw).Return().Once()
	broadcast.Handler("audio").ServeTrack(tw)
	assert.NotNil(t, broadcast.Handler("missing"))
}

func TestBroadcastRemoveTrack_RemovesCatalogEntryAndHandler(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)

	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })
	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.NoError(t, err)

	removed := broadcast.RemoveTrack("video")
	assert.True(t, removed)
	assert.Empty(t, broadcast.Catalog().Tracks)
	_, ok := broadcast.Handler("video").(*MockTrackHandler)
	assert.False(t, ok)
}

func TestBroadcastCatalogBytes_EncodesCurrentCatalog(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)
	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })

	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.NoError(t, err)

	data, err := broadcast.CatalogBytes()
	require.NoError(t, err)

	var decoded Catalog
	require.NoError(t, json.Unmarshal(data, &decoded))
	require.Len(t, decoded.Tracks, 1)
	assert.Equal(t, "video", decoded.Tracks[0].Name)
}

func TestBroadcastRegisterTrack_RejectsReservedCatalogTrackName(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)
	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })

	err = broadcast.RegisterTrack(Track{
		Name:      string(DefaultCatalogTrackName),
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "reserved")
}

func TestBroadcastRegisterTrack_RejectsInvalidInput(t *testing.T) {
	tests := map[string]struct {
		track        Track
		handler      moqt.TrackHandler
		errorMessage string
	}{
		"nil handler": {
			track:        Track{Name: "video", Packaging: PackagingLOC, IsLive: new(false)},
			handler:      nil,
			errorMessage: "track handler cannot be nil",
		},
		"nil handler func": {
			track:        Track{Name: "video", Packaging: PackagingLOC, IsLive: new(false)},
			handler:      moqt.TrackHandlerFunc(nil),
			errorMessage: "track handler function cannot be nil",
		},
		"empty track name": {
			track:        Track{Packaging: PackagingLOC, IsLive: new(false)},
			handler:      &MockTrackHandler{},
			errorMessage: "track name is required",
		},
	}

	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			broadcast, err := NewBroadcast(Catalog{Version: 1})
			require.NoError(t, err)

			err = broadcast.RegisterTrack(tt.track, tt.handler)
			require.Error(t, err)
			assert.Contains(t, err.Error(), tt.errorMessage)
		})
	}
}

func TestBroadcastRegisterTrack_ReplacesExistingTrackAndHandler(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)

	firstHandler := &MockTrackHandler{}
	secondHandler := &MockTrackHandler{}
	t.Cleanup(func() { firstHandler.AssertExpectations(t) })
	t.Cleanup(func() { secondHandler.AssertExpectations(t) })

	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(true),
		Label:     "primary",
	}, firstHandler)
	require.NoError(t, err)

	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(false),
		Label:     "replacement",
	}, secondHandler)
	require.NoError(t, err)

	catalog := broadcast.Catalog()
	require.Len(t, catalog.Tracks, 1)
	assert.Equal(t, "replacement", catalog.Tracks[0].Label)
	require.NotNil(t, catalog.Tracks[0].IsLive)
	assert.False(t, *catalog.Tracks[0].IsLive)
	tw := &moqt.TrackWriter{TrackName: "video"}
	secondHandler.On("ServeTrack", tw).Return().Once()
	broadcast.Handler("video").ServeTrack(tw)
}

func TestBroadcastZeroValue_UsesDefaultCatalogTrackName(t *testing.T) {
	var broadcast Broadcast

	assert.Equal(t, DefaultCatalogTrackName, broadcast.CatalogTrackName())
	assert.NotNil(t, broadcast.Handler(DefaultCatalogTrackName))
	assert.Zero(t, broadcast.Catalog().Version)
	assert.Empty(t, broadcast.Catalog().Tracks)

	require.NoError(t, broadcast.SetCatalog(Catalog{Version: 1}))
	assert.Equal(t, DefaultCatalogTrackName, broadcast.CatalogTrackName())
	assert.Equal(t, 1, broadcast.Catalog().Version)
	assert.NotNil(t, broadcast.Handler(DefaultCatalogTrackName))
	assert.False(t, broadcast.RemoveTrack(DefaultCatalogTrackName))
	assert.False(t, broadcast.RemoveTrack(""))
	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })
	assert.Error(t, broadcast.RegisterTrack(Track{Name: string(DefaultCatalogTrackName)}, handler))
}

func TestBroadcastSetCatalog_PrunesRemovedTrackHandlers(t *testing.T) {
	broadcast, err := NewBroadcast(Catalog{Version: 1})
	require.NoError(t, err)

	handler := &MockTrackHandler{}
	t.Cleanup(func() { handler.AssertExpectations(t) })
	err = broadcast.RegisterTrack(Track{
		Name:      "video",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}, handler)
	require.NoError(t, err)

	err = broadcast.SetCatalog(Catalog{Version: 1, Tracks: []Track{{
		Name:      "audio",
		Packaging: PackagingLOC,
		IsLive:    new(false),
	}}})
	require.NoError(t, err)

	assert.NotNil(t, broadcast.Handler("audio"))
	_, ok := broadcast.Handler("video").(*MockTrackHandler)
	assert.False(t, ok)
}

File: moq-web/src/msf/broadcast_test.ts
import { assertEquals, assertRejects, assertThrows } from "@std/assert";
import { Broadcast, DefaultCatalogTrackName } from "./mod.ts";
import type { TrackHandler } from "../track_mux.ts";
import type { TrackWriter } from "../track_writer.ts";

class MockTrackHandler implements TrackHandler {
	calls: TrackWriter[] = [];

	async serveTrack(trackWriter: TrackWriter): Promise<void> {
		this.calls.push(trackWriter);
	}
}

function newCatalogTrackWriter(trackName: string) {
	const frames: Uint8Array[] = [];
	const closeWithErrorCalls: number[] = [];
	let closeCalls = 0;
	let groupCloseCalls = 0;
	const group = {
		async writeFrame(payload: Uint8Array): Promise<Error | undefined> {
			frames.push(payload);
			return undefined;
		},
		async close(): Promise<void> {
			groupCloseCalls++;
		},
		async cancel(): Promise<void> {
			groupCloseCalls++;
		},
	};
	const trackWriter = {
		trackName,
		async openGroup() {
			return [group, undefined] as const;
		},
		async closeWithError(code: number): Promise<void> {
			closeWithErrorCalls.push(code);
		},
		async close(): Promise<void> {
			closeCalls++;
		},
	} as unknown as TrackWriter & {
		frames: Uint8Array[];
		closeWithErrorCalls: number[];
		getCloseCalls: () => number;
		getGroupCloseCalls: () => number;
	};
	trackWriter.frames = frames;
	trackWriter.closeWithErrorCalls = closeWithErrorCalls;
	trackWriter.getCloseCalls = () => closeCalls;
	trackWriter.getGroupCloseCalls = () => groupCloseCalls;
	return trackWriter;
}

Deno.test("msf Broadcast registerTrack updates catalog and handler", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const handler = new MockTrackHandler();

	await broadcast.registerTrack(
		{ name: "video", packaging: "loc", isLive: false },
		handler,
	);

	assertEquals(broadcast.catalog().tracks.length, 1);
	assertEquals(broadcast.catalog().tracks[0]?.name, "video");

	const trackWriter = { trackName: "video" } as TrackWriter;
	await broadcast.handler("video").serveTrack(trackWriter);
	assertEquals(handler.calls.length, 1);
	assertEquals(handler.calls[0], trackWriter);
});

Deno.test("msf Broadcast serves catalog on reserved track", async () => {
	const broadcast = new Broadcast({
		version: 1,
		tracks: [{ name: "video", packaging: "loc", isLive: false }],
	});
	const trackWriter = newCatalogTrackWriter(DefaultCatalogTrackName);

	await broadcast.serveTrack(trackWriter);

	assertEquals(trackWriter.frames.length, 1);
	const text = new TextDecoder().decode(trackWriter.frames[0]);
	assertEquals(text.includes('"video"'), true);
	assertEquals(trackWriter.getGroupCloseCalls(), 1);
	assertEquals(trackWriter.getCloseCalls(), 1);
	assertEquals(trackWriter.closeWithErrorCalls.length, 0);
});

Deno.test("msf Broadcast rejects duplicate track names across namespaces", () => {
	assertThrows(
		() =>
			new Broadcast({
				version: 1,
				defaultNamespace: "live/main",
				tracks: [
					{ name: "video", packaging: "loc", isLive: false },
					{
						namespace: "live/backup",
						name: "video",
						packaging: "loc",
						isLive: false,
					},
				],
			}),
		Error,
		"unique track names across namespaces",
	);
});

Deno.test("msf Broadcast catalogTrackName returns custom name", () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] }, "meta");
	assertEquals(broadcast.catalogTrackName, "meta");
});

Deno.test("msf Broadcast registerTrack rejects nil handler", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	await assertRejects(
		() =>
			broadcast.registerTrack(
				{ name: "video", packaging: "cmaf" },
				null as unknown as TrackHandler,
			),
		Error,
		"cannot be nil",
	);
});

Deno.test("msf Broadcast registerTrack rejects missing track name", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const handler = new MockTrackHandler();
	await assertRejects(
		() => broadcast.registerTrack({ name: "", packaging: "cmaf" }, handler),
		Error,
		"track name is required",
	);
});

Deno.test("msf Broadcast registerTrack rejects catalog track name", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const handler = new MockTrackHandler();
	await assertRejects(
		() =>
			broadcast.registerTrack({ name: DefaultCatalogTrackName, packaging: "cmaf" }, handler),
		Error,
		"reserved for the catalog track",
	);
});

Deno.test("msf Broadcast registerTrack replaces existing track", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [{ name: "audio", packaging: "cmaf" }] });
	const handler1 = new MockTrackHandler();
	await broadcast.registerTrack({ name: "audio", packaging: "cmaf", bitrate: 128 }, handler1);
	assertEquals(broadcast.catalog().tracks.length, 1);
	assertEquals(broadcast.catalog().tracks[0]?.bitrate, 128);
});

Deno.test("msf Broadcast setCatalog removes stale tracks", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [{ name: "video", packaging: "cmaf" }] });
	const handler = new MockTrackHandler();
	await broadcast.registerTrack({ name: "video", packaging: "cmaf" }, handler);

	await broadcast.setCatalog({ version: 1, tracks: [{ name: "audio", packaging: "cmaf" }] });
	assertEquals(broadcast.catalog().tracks.length, 1);
	assertEquals(broadcast.catalog().tracks[0]?.name, "audio");
});

Deno.test("msf Broadcast setCatalog with empty previous tracks", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	await broadcast.setCatalog({ version: 1, tracks: [{ name: "video", packaging: "cmaf" }] });
	assertEquals(broadcast.catalog().tracks.length, 1);
});

Deno.test("msf Broadcast removeTrack removes registered track", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [{ name: "video", packaging: "cmaf" }] });
	const handler = new MockTrackHandler();
	await broadcast.registerTrack({ name: "video", packaging: "cmaf" }, handler);

	const removed = await broadcast.removeTrack("video");
	assertEquals(removed, true);
	assertEquals(broadcast.catalog().tracks.length, 0);
});

Deno.test("msf Broadcast removeTrack returns false for empty name", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const removed = await broadcast.removeTrack("");
	assertEquals(removed, false);
});

Deno.test("msf Broadcast removeTrack returns false for catalog track name", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const removed = await broadcast.removeTrack(DefaultCatalogTrackName);
	assertEquals(removed, false);
});

Deno.test("msf Broadcast removeTrack returns true when track not in catalog but was registered", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	const handler = new MockTrackHandler();
	await broadcast.registerTrack({ name: "audio", packaging: "cmaf" }, handler);

	// Remove from catalog first via setCatalog, then removeTrack
	const removed = await broadcast.removeTrack("nonexistent");
	assertEquals(removed, false);
});

Deno.test("msf Broadcast close does not throw", async () => {
	const broadcast = new Broadcast({ version: 1, tracks: [] });
	await broadcast.close();
});

Deno.test("msf Broadcast serveCatalogTrack handles openGroup error", async () => {
	const broadcast = new Broadcast({
		version: 1,
		tracks: [{ name: "video", packaging: "cmaf" }],
	});
	const closeWithErrorCalls: number[] = [];
	const trackWriter = {
		trackName: DefaultCatalogTrackName,
		async openGroup() {
			return [undefined, new Error("open failed")] as const;
		},
		async closeWithError(code: number): Promise<void> {
			closeWithErrorCalls.push(code);
		},
		async close(): Promise<void> {},
	} as unknown as TrackWriter;

	await broadcast.serveTrack(trackWriter);
	assertEquals(closeWithErrorCalls.length, 1);
});

Deno.test("msf Broadcast serveCatalogTrack handles writeFrame error", async () => {
	const broadcast = new Broadcast({
		version: 1,
		tracks: [{ name: "video", packaging: "cmaf" }],
	});
	const closeWithErrorCalls: number[] = [];
	const cancelCalls: number[] = [];
	const group = {
		async writeFrame(_payload: Uint8Array): Promise<Error> {
			return new Error("write failed");
		},
		async cancel(code: number): Promise<void> {
			cancelCalls.push(code);
		},
		async close(): Promise<void> {},
	};
	const trackWriter = {
		trackName: DefaultCatalogTrackName,
		async openGroup() {
			return [group, undefined] as const;
		},
		async closeWithError(code: number): Promise<void> {
			closeWithErrorCalls.push(code);
		},
		async close(): Promise<void> {},
	} as unknown as TrackWriter;

	await broadcast.serveTrack(trackWriter);
	assertEquals(cancelCalls.length, 1);
	assertEquals(closeWithErrorCalls.length, 1);
});
